### ISSUE TITLE
Sort: on the statehist table silently corrupts all returned rows (no crash, wrong data)

### ISSUE THREAD

**Issue by kfiresmith** [naemon/naemon-livestatus#140]
Environment

naemon-livestatus: 1.5.1 (1.5.1.source-naemon per GET status)
naemon-core: 1.5.2
OS: Ubuntu 26.04 LTS
Installed from: apt packages (naemon-livestatus, naemon-core, libnaemon 1.5.2/1.5.1)
Summary

Adding Sort: from asc (or any Sort: header) to a GET statehist query causes every row in the response to come back with the same values — specifically, every row's columns get overwritten with the last row's data. This is silent data corruption, not a crash: the request succeeds, returns the correct row count, but every row is identical (and wrong) instead of distinct.

This is different from the known sort-related crash reports on this tracker (#39 SIGABRT on a services sort, #70 CentOS 8 assertion failure on a hosts sort) — no crash occurs here at all, and it's isolated to statehist.

Steps to reproduce

Query statehist twice for the same host/time range, once without Sort: and once with it, over the raw livestatus socket (e.g. /var/cache/naemon/live):

Without Sort: (correct):

GET statehist
Columns: from until duration state in_downtime log_output
Filter: host_name = localhost
Filter: service_description =
Filter: time >= 1787917855
Filter: time <= 1788522655
OutputFormat: json
Returns 9 distinct rows, e.g.:

[1787917855, 1788485190, 567335, -1, 0, '']
[1788485190, 1788487899, 2709, 1, 0, '(No output on stdout) stderr: ...']
[1788487899, 1788518692, 30793, 0, 0, 'PING OK - Packet loss = 0%, RTA = 0.06 ms']
...
[1788520011, 1788522655, 2644, 0, 0, '']
With Sort: from asc added (everything else identical):

GET statehist
Columns: from until duration state in_downtime log_output
Filter: host_name = localhost
Filter: service_description =
Filter: time >= 1787917855
Filter: time <= 1788522655
Sort: from asc
OutputFormat: json
Still returns 9 rows (correct count), but every row is now identical to the last row from the unsorted query:

[1788522655, 1788522655, 2644, 0, 0, '']
[1788522655, 1788522655, 2644, 0, 0, '']
[1788522655, 1788522655, 2644, 0, 0, '']
... (9 times)
Expected: the 9 rows returned in ascending order by from, with their real distinct values.

Actual: 9 rows, correct count, but every row's data replaced with the last row's values.

Notes

Reproduced with a raw client speaking LQL directly over the Unix socket — not specific to any particular client library, wrapper, or language.
Sort: works correctly on other tables we tested (e.g. log) — the bug appears isolated to statehist.
Workaround: omit Sort: from statehist queries and sort the returned rows client-side instead.

DISCLAIMER:  Bug report drafted with assistance from Claude Sonnet 5 high, but I'm a real human working on the project to switch to naemon after having been a Nagios user for like 20 years and happy to participate in any future testing, data gathering, corrections to our report, etc.  Hopefully that's OK!

**Comment by sni** [comment]
thanks for your report, it should be fixed in the next release.
